serverless/internal/storage/fs/fs.go

```diff
@@ -0,0 +1,129 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package fs provides a simple filesystem log storage implementation.
+package fs
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/google/trillian-examples/serverless/api"
+	"github.com/google/trillian-examples/serverless/internal/storage/fs/layout"
+)
+
+// Storage is a serverless storage implementation which uses files to store tree state.
+// The on-disk structure is:
+//  <rootDir>/leaves/aa/bb/cc/ddeeff...
+//  <rootDir>/leaves/pending/
```

Comment:
Whoops, this slipped through, it was supposed to appear later! (removed for now)
Yeah, this won't fully stop dupes in the log of course, but can help to avoid pending dupes.